### PR TITLE
Add utils subroutine to handle interactive command

### PR DESCRIPTION
**Pull request by wnereiz**
It mainly for security tools testing like AppArmor or future secure boot testing.

It involved using typescript to disable pipe buffer, so that avoid the 'dirty hack' on the env or code. And we do not need to make it different for legacy application (writing by perl) or the newly one (by python).

There are several verification runs for this function used on AppArmor cases to replace old inelegant approach. But they are not pushed yet. I plan to submit them after this PR merged.

- Related ticket: https://progress.opensuse.org/issues/45785
- Verification run:
   - SLE12SP4: http://10.67.17.9/tests/218#step/aa_logprof/22
   - SLE15SP1: http://10.67.17.9/tests/217#step/aa_logprof/22
   - Tumbleweed: http://10.67.17.9/tests/220#step/aa_logprof/22
   - mokutil: http://10.67.17.9/tests/221#step/mokutil_sign/18

### CODE REVIEW

**Comment by wnereiz** [comment]
Format of some places have been modified to satisfy Travis

File: lib/utils.pm [hunk 1]
```diff
@@ -163,19 +164,23 @@ sub integration_services_check_ip {
     my $vmname = console('svirt')->name;
     my $ips_host_pov;
     if (check_var('VIRSH_VMM_FAMILY', 'hyperv')) {
-        $ips_host_pov = console('svirt')->get_cmd_output(
-            'powershell -Command "Get-VM ' . $vmname . ' | Get-VMNetworkAdapter | Format-Table -HideTableHeaders IPAddresses"');
+        $ips_host_pov = console('svirt')
+          ->get_cmd_output('powershell -Command "Get-VM ' . $vmname . ' | Get-VMNetworkAdapter | Format-Table -HideTableHeaders IPAddresses"');
```
Comment (okurz) [inline lib/utils.pm]: these changes do not seem to relate to your commit. I suggest to revert them or keep them for another PR with a dedicated description.

Comment (wnereiz) [reply lib/utils.pm]: I did these changes to satisfy the travis tidy process. Do you have a better suggestion to make travis checking pass and do not need to touch these codes?

Comment (wnereiz) [reply lib/utils.pm]: **expect** is not used here, because we want more control with the procedure during the interactive. With expect command, all the interactive will be handled by the command itself, there's no way to control it from test script level. 

It is not a big problem currently, but in future if we want to extend the function, eg. make additional action according to some certain prompt message, it will be difficult. And as an addition, we also want to use _type_string_ and _send_key_ explicitly according to test case develper's choice.

**Comment by wnereiz** [comment]
Thank you for the useful information about commit message. I made a change here.

**Comment by wnereiz** [comment]
**expect** is not used here, because we want more control with the procedure during the interactive. With expect command, all the interactive will be handled by the command itself, there's no way to control it from test script level. 

It is not a big problem currently, but in future if we want to extend the function, eg. make additional action according to some certain prompt message, it will be difficult. And as an addition, we also want to use _type_string_ and _send_key_ explicitly according to test case develper's choice.

File: lib/utils.pm [hunk 2]
```diff
@@ -1120,4 +1121,54 @@ sub prepare_ssh_localhost_key_login {
...
+                        type_string $i->{string};
+                        last;
+                    }
+                    elsif ($i->{key}) {
+                        send_key $i->{key};
+                        last;
+                    }
+                    else {
+                        die "$i->{prompt} - No flags specified";
+                    }
```
Comment (okurz) [inline lib/utils.pm]: Could you try to reduce the complexity/indendation within this loop?

File: lib/utils.pm [hunk 3]
```diff
@@ -1120,4 +1121,54 @@ sub prepare_ssh_localhost_key_login {
     }
 }
 
+=head2 script_run_interactive
+For interactive command, input strings or keys according to the prompt message
+in the run time. Pass arrayref which contains the prompt message to be matched
+(regex) and the answers
```
Comment (rwx788) [inline lib/utils.pm]: Could you please add an example of the usage as you use e.g. hard-coded `$endmark`, which one can observe on serial during the investigation.

File: lib/utils.pm [hunk 4]
```diff
@@ -1120,4 +1121,54 @@ sub prepare_ssh_localhost_key_login {
...
+    }
+
+    for my $k (@$scan) {
+        push(@words, $k->{prompt});
+    }
+
+    my $endmark = "Script done.*\/dev\/null";
+    push(@words, $endmark);
+
+  LOOP: {
```
Comment (rwx788) [inline lib/utils.pm]: I'm not sure if this label helps in anyway, as everyone can see that it's a loop below. So I would either rename it that it explains what the loop does or get rid of it. You could also simply wrap it in the subroutine.

File: lib/utils.pm (hunk 2 above)
Comment (rwx788) [reply lib/utils.pm]: For example you could write
```perl
next if ($output !~ $i->{prompt});
``` 
to reduce nested level by one.

Comment (wnereiz) [reply lib/utils.pm]: @okurz @rwx788  Thank you for the great suggestion, I learned a lot from you. Changes have been made. Please check it.

File: lib/utils.pm (hunk 3 above)
Comment (wnereiz) [reply lib/utils.pm]: I added an example for calling this subroutine, as well as a explanation for the mark of command finishing.

File: lib/utils.pm (hunk 4 above)
Comment (wnereiz) [reply lib/utils.pm]: Removed label which is not meaningful as you described

**Comment by wnereiz** [comment]
@rwx788 squashed. 

> Have you considered moving to `os-autoinst`? Code looks fine to me....

Yes, it would be great that we could move it to os-autoinst as an API. But before that I would like to run it in lib/utils.pm for some time, and collect some feedback. Now I hope it could be merged so that we can use it for following test cases about mokutil

**Comment by rwx788** [comment]
@wnereiz looking forward to the outcomes, looks helpful for many scenarios.

File: lib/utils.pm (hunk 2 above)
Comment (okurz) [reply lib/utils.pm]: Sorry if I could not come up with a more helpful, specific suggestion. Thanks to @rwx788 for the hint.
